### PR TITLE
PixelPaint: Move tool scaling function now works in all directions

### PR DESCRIPTION
**Pull request by tcl3**
It is now possible to scale the current layer using the move tool from all four corners of the layer boundary. Previously scaling was only possible from the bottom right of the image.

This patch also clamps the width and height of the scaled image to 1 pixel, rather than growing the image once the user scales past the opposite corner.

Here is a video demonstrating the new functionality:

https://user-images.githubusercontent.com/2817754/208587395-06e6afe9-b075-4d44-8ad6-9cf668ef4e5f.mp4

### CODE REVIEW

File: Userland/Applications/PixelPaint/Tools/MoveTool.cpp [hunk 1]
```diff
@@ -49,37 +49,75 @@ void MoveTool::on_mousemove(Layer* layer, MouseEvent& event)
...
+            m_editor->update_tool_cursor();
+        }
     }
 
     if (!m_layer_being_moved)
         return;
 
+    auto cursor_position = event.image_event().position();
+    auto delta = cursor_position - m_event_origin;
+    int x;
```
Comment (martinfalisse) [inline Userland/Applications/PixelPaint/Tools/MoveTool.cpp:65]: You could use an IntRect for these variables x, y, width, and height. Perhaps you could assign directly to the m_new_layer_rect ?

File: Userland/Applications/PixelPaint/Tools/MoveTool.h [hunk 2]
```diff
@@ -27,17 +33,16 @@ class MoveTool final : public Tool {
...
     Gfx::IntPoint m_event_origin;
     Gfx::IntPoint m_layer_origin;
-    Gfx::IntPoint m_new_scaled_layer_location;
-    Gfx::IntSize m_new_layer_size;
+    Gfx::IntRect m_new_layer_rect;
     bool m_scaling { false };
-    bool m_mouse_in_resize_corner { false };
-    bool m_keep_ascept_ratio { false };
-
+    Optional<Corner const> m_mouse_in_resize_corner {};
```
Comment (martinfalisse) [inline Userland/Applications/PixelPaint/Tools/MoveTool.h:44]: This variable name seems awkward to me. Maybe m_resize_corner_of_mouse, or maybe you have an even better variable name idea :^)

File: Userland/Applications/PixelPaint/Tools/MoveTool.h [hunk 3]
```diff
@@ -27,17 +33,16 @@ class MoveTool final : public Tool {
 
 private:
     virtual StringView tool_name() const override { return "Move Tool"sv; }
-    ErrorOr<void> update_cached_preview_bitmap(Layer const* layer);
+    ErrorOr<void> update_cached_preview_bitmap(Layer const*);
+    Optional<Corner const> mouse_in_resize_corner(Layer const*, MouseEvent&);
```
Comment (martinfalisse) [inline Userland/Applications/PixelPaint/Tools/MoveTool.h:37]: Same, I would try to find a better name for this function. When I read it I think it returns a boolean

File: Userland/Applications/PixelPaint/Tools/MoveTool.h [hunk 4]
```diff
@@ -27,17 +33,16 @@ class MoveTool final : public Tool {
 
 private:
     virtual StringView tool_name() const override { return "Move Tool"sv; }
-    ErrorOr<void> update_cached_preview_bitmap(Layer const* layer);
+    ErrorOr<void> update_cached_preview_bitmap(Layer const*);
```
Comment (martinfalisse) [inline Userland/Applications/PixelPaint/Tools/MoveTool.h:36]: I think this is kind of an unnecessary change

File: Userland/Applications/PixelPaint/Tools/MoveTool.h [hunk 5]
```diff
@@ -12,6 +12,12 @@
 
 namespace PixelPaint {
 
+enum class Corner {
+    TopLeft,
+    TopRight,
```
Comment (martinfalisse) [inline Userland/Applications/PixelPaint/Tools/MoveTool.h:17]: I'm wondering if in the codebase enums tend to be alphabetized or not

File: Userland/Applications/PixelPaint/Tools/MoveTool.cpp [hunk 6]
```diff
@@ -101,21 +139,22 @@ void MoveTool::on_mouseup(Layer* layer, MouseEvent& event)
...
+    m_new_layer_rect = m_editor->active_layer()->rect();
     m_editor->update_tool_cursor();
     m_editor->did_complete_action(tool_name());
 }
 
 bool MoveTool::on_keydown(GUI::KeyEvent& event)
 {
     if (event.key() == Key_Shift)
-        m_keep_ascept_ratio = true;
+        m_keep_aspect_ratio = true;
```
Comment (martinfalisse) [inline Userland/Applications/PixelPaint/Tools/MoveTool.cpp:157]: I think this should be in a separate commit

File: Userland/Applications/PixelPaint/Tools/MoveTool.cpp (hunk 1 above)
Comment (tcl3) [reply Userland/Applications/PixelPaint/Tools/MoveTool.cpp]: updated to assign directly to `m_new_layer_rect`

File: Userland/Applications/PixelPaint/Tools/MoveTool.h (hunk 2 above)
Comment (tcl3) [reply Userland/Applications/PixelPaint/Tools/MoveTool.h]: I changed the `Corner` enum name to `ResizeAnchorLocation` and changed `m_mouse_in_resize_corner` to `m_resize_anchor_location`

File: Userland/Applications/PixelPaint/Tools/MoveTool.h (hunk 3 above)
Comment (tcl3) [reply Userland/Applications/PixelPaint/Tools/MoveTool.h]: Changed to `resize_anchor_location_from_cursor_position`. I don't love it, as its a bit verbose, but at least its descriptive.

File: Userland/Applications/PixelPaint/Tools/MoveTool.cpp (hunk 6 above)
Comment (tcl3) [reply Userland/Applications/PixelPaint/Tools/MoveTool.cpp:149]: split into a separate commit

**Comment by tcl3** [comment]
@martinfalisse I have applied your suggestions. I also found an issue when scaling with a fixed aspect ratio from the top right and bottom left, which I have fixed.

**Review by martinfalisse — COMMENTED** [review]
Nice, I think it's much more legible now.

You may have noticed but the new layer after scaling doesn't seem to align correctly with the corners of the canvas. Do you know what this is caused by? Maybe it's caused by some int conversion.

Edit: for example in the video at 0:11. It's still reproducible in the latest changes as far as I can tell.

File: Userland/Applications/PixelPaint/Tools/MoveTool.h [hunk 7]
```diff
@@ -28,16 +34,15 @@ class MoveTool final : public Tool {
...
     Gfx::IntPoint m_layer_origin;
-    Gfx::IntPoint m_new_scaled_layer_location;
-    Gfx::IntSize m_new_layer_size;
+    Gfx::IntRect m_new_layer_rect;
     bool m_scaling { false };
-    bool m_mouse_in_resize_corner { false };
+    Optional<ResizeAnchorLocation const> m_resize_anchor_location {};
     bool m_keep_ascept_ratio { false };
-
     RefPtr<Gfx::Bitmap> m_cached_preview_bitmap { nullptr };
```
[the remark this answers is not in the document]
Comment (tcl3) [reply Userland/Applications/PixelPaint/Tools/MoveTool.h:47]: reverted

**Comment by martinfalisse** [comment]
FYI the first commit doesn't build anymore as the m_keep_aspect_ratio correction is split up between commits.

**Comment by tcl3** [comment]
You may have noticed but the new layer after scaling doesn't seem to align correctly with the corners of the canvas. Do you know what this is caused by? Maybe it's caused by some int conversion.
> 
> Edit: for example in the video at 0:11. It's still reproducible in the latest changes as far as I can tell.

So I've done some investigation of this. I believe that there can be some error introduced when converting from frame to image co-ordinates. The fact that I was relying on `image_event.position()` to  calculate the coordinate of the opposite corner is what led to the issue.

I have rewritten the code in what I believe is a much much understandable way. The new version ensures the opposite corner always maintains it's position. This ensures that the far edge of the image doesn't jump when you finish scaling. 

I have not reimplemented the part of the commit which clamps the size of the image to 1 pixel if the user scales past the edge. Other tools don't do this and I couldn't find a way to do it without doubling the size of the code.

I also found an issue where you could crash the program by clicking on the resize anchor without moving it. This issue was also present in the master branch. I've wasn't sure whether to separate the fix out or not, as the line that fixes it is needed in this commit anyway.

There is still an issue which sometimes causes a slight white border to show up around the edges of the image. This is a painting issue. If you zoom in you can see that all the pixels are where they are supposed to be

Here's a video of the new changes showing a more stable opposite edge:

https://user-images.githubusercontent.com/2817754/208851914-ccc96de8-414b-48f5-95ac-d61080dd68a1.mp4
